File: java/lib/src/main/java/io/openweb3/xwebhook/Statistics.java
package io.openweb3.xwebhook;

import io.openweb3.xwebhook.exceptions.ApiException;
import io.openweb3.xwebhook.internal.api.StatisticsApi;
import io.openweb3.xwebhook.models.AggregateEventTypesOut;
import io.openweb3.xwebhook.models.AppUsageStatsIn;
import io.openweb3.xwebhook.models.AppUsageStatsOut;

public class Statistics {
    private final StatisticsApi api;

    Statistics() {
        api = new StatisticsApi();
    }

    public AppUsageStatsOut aggregateAppStats(final AppUsageStatsIn appUsageStatsIn, final PostOptions options) throws ApiException {
        try {
            return api.v1StatisticsAggregateAppStats(appUsageStatsIn, options.getIdempotencyKey());
        } catch (io.openweb3.xwebhook.internal.ApiException e) {
            throw Utils.wrapInternalApiException(e);
        }
    }

    public AggregateEventTypesOut aggregateEventTypes() throws io.openweb3.xwebhook.internal.ApiException {
        try {
            return api.v1StatisticsAggregateEventTypes();
        } catch (io.openweb3.xwebhook.internal.ApiException e) {
            throw Utils.wrapInternalApiException(e);
        }
    }

}

File: kotlin/lib/src/main/kotlin/Statistics.kt
package io.openweb3.xwebhook.kotlin

import io.openweb3.xwebhook.kotlin.exceptions.ApiException
import io.openweb3.xwebhook.kotlin.internal.apis.StatisticsApi
import io.openweb3.xwebhook.kotlin.models.AggregateEventTypesOut
import io.openweb3.xwebhook.kotlin.models.AppUsageStatsIn
import io.openweb3.xwebhook.kotlin.models.AppUsageStatsOut

class Statistics internal constructor(token: String, options: XwebhookOptions) {
    private val api = StatisticsApi(options.serverUrl)

    init {
        api.accessToken = token
        api.userAgent = options.getUA()
        options.initialRetryDelayMillis?.let { api.initialRetryDelayMillis = it }
        options.numRetries?.let { api.numRetries = it }
    }

    suspend fun aggregateAppStats(
        appUsageStatsIn: AppUsageStatsIn,
        options: PostOptions = PostOptions(),
    ): AppUsageStatsOut {
        try {
            return api.v1StatisticsAggregateAppStats(appUsageStatsIn, options.idempotencyKey)
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }

    suspend fun aggregateEventTypes(): AggregateEventTypesOut {
        try {
            return api.v1StatisticsAggregateEventTypes()
        } catch (e: Exception) {
            throw ApiException.wrap(e)
        }
    }
}
